clientconn.go

```diff
@@ -248,37 +248,30 @@ func DialContext(ctx context.Context, target string, opts ...DialOption) (conn *
 	}
 
 	// Determine the resolver to use.
-	cc.parsedTarget = grpcutil.ParseTarget(cc.target, cc.dopts.copts.Dialer != nil)
-	channelz.Infof(logger, cc.channelzID, "parsed scheme: %q", cc.parsedTarget.Scheme)
-	resolverBuilder := cc.getResolver(cc.parsedTarget.Scheme)
-	if resolverBuilder == nil {
-		// If resolver builder is still nil, the parsed target's scheme is
-		// not registered. Fallback to default resolver and set Endpoint to
-		// the original target.
-		channelz.Infof(logger, cc.channelzID, "scheme %q not registered, fallback to default scheme", cc.parsedTarget.Scheme)
-		cc.parsedTarget = resolver.Target{
-			Scheme:   resolver.GetDefaultScheme(),
-			Endpoint: target,
-		}
-		resolverBuilder = cc.getResolver(cc.parsedTarget.Scheme)
-		if resolverBuilder == nil {
-			return nil, fmt.Errorf("could not get resolver for default scheme: %q", cc.parsedTarget.Scheme)
-		}
+	resolverBuilder, err := cc.parseTargetAndFindResolver()
+	if err != nil {
+		return nil, err
 	}
 
-	creds := cc.dopts.copts.TransportCredentials
-	if creds != nil && creds.Info().ServerName != "" {
-		cc.authority = creds.Info().ServerName
-	} else if cc.dopts.insecure && cc.dopts.authority != "" {
-		cc.authority = cc.dopts.authority
-	} else if strings.HasPrefix(cc.target, "unix:") || strings.HasPrefix(cc.target, "unix-abstract:") {
+	// Determine the authority for the channel. For now, we will use the parsed
+	// endpoint from the user's dial target (with a leading "/" stripped), as
+	// the default value. We do have some special handling for targets with unix
+	// scheme and empty hostname (":port").
+	// TODO: Define an optional interface on the name resolver to return the
+	// authority given the user's dial target.
+	cc.authority = strings.TrimPrefix(cc.parsedTarget.Endpoint, "/")
+	if strings.HasPrefix(cc.authority, ":") {
+		cc.authority = "localhost" + cc.authority
+	}
+	if cc.parsedTarget.Scheme == "unix" || cc.parsedTarget.Scheme == "unix-abstract" {
 		cc.authority = "localhost"
-	} else if strings.HasPrefix(cc.parsedTarget.Endpoint, ":") {
-		cc.authority = "localhost" + cc.parsedTarget.Endpoint
-	} else {
-		// Use endpoint from "scheme://authority/endpoint" as the default
-		// authority for ClientConn.
-		cc.authority = cc.parsedTarget.Endpoint
+	}
```

Comment:
Can we eliminate this by making sure the addresses output by the unix resolver set `ServerName`  to `localhost`?  If so, let's do that instead..  I would rather avoid as many special cases as possible in generic places.

Reply:
But for the `unix` resolver, the balancer is going to be pick first, which doesn't/shouldn't care about the authority.  Most balancers shouldn't care, right -- just RLS/gRPCLB?

Reply:
Yes, for the special handling of `unix` and `unix-abstract`.  We shouldn't need to do this if we set `ServerName` in those resolvers instead.  Any balancer used with those schemes shouldn't care about the authority.  Only RLS and gRPCLB care about the "channel authority" (so they can use the same one with their backends), and it doesn't make sense to use those with a unix address.